doc-content/enterprise-only/optimizer/ex-nQueens-ref.adoc

```diff
@@ -0,0 +1,164 @@
+
+Below is a correct solution:
+
+.A Correct Solution for the Four Queens Puzzle
+image::UseCasesAndExamples/NQueens/solvedNQueens04.png[align="center"]
+
+All the constraints have been met, so the solution is correct. 
+
+Note that most _n_ queens puzzles have multiple correct solutions.
+We will focus on finding a single correct solution for a given _n_, not on finding the number of possible correct solutions for a given _n_.
+
+
+[[nQueensProblemSize]]
+.Problem Size
+[source,options="nowrap"]
+----
+4queens   has   4 queens with a search space of    256.
+8queens   has   8 queens with a search space of   10^7.
+16queens  has  16 queens with a search space of  10^19.
+32queens  has  32 queens with a search space of  10^48.
+64queens  has  64 queens with a search space of 10^115.
+256queens has 256 queens with a search space of 10^616.
+----
+
+The implementation of the _n_ queens example has not been optimized because it functions as a beginner example. Nevertheless, it can easily handle 64 queens.
+With a few changes it has been shown to easily handle 5000 queens and more.
+
+
+[[nQueensDomainModel]]
+.Domain Model
+
+This example uses the domain model to solve the four queens problem.
+
+. *Creating a Domain Model* 
+A good domain model will make it easier to understand and solve your planning problem.
++
+This is the domain model for the _n_ queens example:
++
+[source,java,options="nowrap"]
+----
+public class Column {
+    
+    private int index;
+
+    // ... getters and setters
+}
+----
++
+[source,java,options="nowrap"]
+----
+public class Row {
+    
+    private int index;
+
+    // ... getters and setters
+}
+----
++
+[source,java,options="nowrap"]
+----
+public class Queen {
+    
+    private Column column;
+    private Row row;
+
+    public int getAscendingDiagonalIndex() {...}
+    public int getDescendingDiagonalIndex() {...}
+
+    // ... getters and setters
+}
+----
+
+
+. *Calculating the Search Space.*
++
+A `Queen` instance has a `Column` (for example: 0 is column A, 1 is column B, ...) and a `Row` (its row, for example: 0 is row 0, 1 is row 1, ...). 
++
+The ascending diagonal line and the descending diagonal line can be calculated based on the column and the row.
++
+The column and row indexes start from the upper left corner of the chessboard.
++
+[source,java,options="nowrap"]
+----
+public class NQueens {
+    
+    private int n;
+    private List<Column> columnList;
+    private List<Row> rowList;
+
+    private List<Queen> queenList;
+
+    private SimpleScore score;
+
+    // ... getters and setters
+}
+----
++
+. *Finding the Solution*
++
+A single `NQueens` instance contains a list of all `Queen` instances.
+It is the `Solution` implementation which will be supplied to, solved by, and retrieved from the Solver.
+
+Notice that in the four queens example, NQueens's `getN()` method will always return four.
+
+.A Solution for Four Queens Shown in the Domain Model
```

Comment:
The table is nearly unusable in the output because there are too many columns and the last two columns have long, crazily wrapped heads. I'd remove the "A solution" column and content and make it stand alone, followed by the now 5-columned table that's at least readable now.
```
.A solution for four ...
[image]

.Solution details
[cols="6a,1,1,1,1", options="header"]
 |===
 | Queen
 | columnIndex
 | rowIndex
 | ascendingDiagonalIndex (columnIndex + rowIndex)
 | descendingDiagonalIndex (columnIndex - rowIndex)

etc.
.....
```